src/local-mode.ts

```diff
@@ -56,42 +90,172 @@ export class LocalModeManager {
+		const volumes = this.docker
+			.listVolumes()
+			.then(resp => _.map(resp.Volumes, 'Name'));
+		const networks = this.docker.listNetworks().then(resp => _.map(resp, 'Id'));
+
+		return Bluebird.all([containers, images, volumes, networks]).then(
+			data =>
+				new EngineSnapshotRecord(
+					new EngineSnapshot(data[0], data[1], data[2], data[3]),
+					new Date(),
+				),
+		);
+	}
+
+	// Store engine snapshot data in the local database.
+	public async storeEngineSnapshot(
+		record: EngineSnapshotRecord,
+	): Promise<void> {
+		const timestamp = record.timestamp.toISOString();
+		log.info(
+			`Storing engine snapshot in the database. Timestamp: ${timestamp}`,
+		);
+		return this.db
+			.models('engineSnapshot')
+			.delete()
+			.then(() => {
+				return this.db.models('engineSnapshot').insert({
+					snapshot: JSON.stringify(record.snapshot),
+					timestamp,
+				});
+			});
+	}
 
-			await Bluebird.map(containers, containerId => {
-				log.debug('Removing local mode container: ', containerId);
-				return this.docker.getContainer(containerId).remove({ force: true });
+	// Read the latest stored snapshot from the database.
+	public async retrieveLatestSnapshot(): Promise<EngineSnapshotRecord | null> {
+		return this.db
+			.models('engineSnapshot')
+			.select()
+			.orderBy('rowid', 'DESC')
+			.first()
+			.then(r => {
+				if (!r) {
+					return null;
+				}
+				return new EngineSnapshotRecord(
+					EngineSnapshot.fromJSON(r.snapshot),
+					new Date(Date.parse(r.timestamp)),
+				);
 			});
-			await Bluebird.map(images, imageId => {
-				log.debug('Removing local mode image: ', imageId);
-				return this.docker.getImage(imageId).remove({ force: true });
+	}
+
+	private async removeLocalModeArtifacts(
+		objects: EngineSnapshot,
+	): Promise<void> {
+		log.debug(`Going to delete following objects: ${objects}`);
+
+		const map = Bluebird.map;
+
+		return map(objects.containers, cId => {
+			return this.docker
+				.getContainer(cId)
+				.remove({ force: true })
+				.catch(e => log.error(`Unable to delete container ${cId}`, e));
+		})
+			.then(() =>
+				map(objects.images, iId => {
+					return this.docker
+						.getImage(iId)
+						.remove({ force: true })
+						.catch(e => log.error(`Unable to delete image ${iId}`, e));
+				}),
+			)
+			.then(() =>
+				map(objects.networks, nId => {
+					return this.docker
+						.getNetwork(nId)
+						.remove()
+						.catch(e => log.error(`Unable to delete network ${nId}`, e));
+				}),
+			)
+			.then(() =>
+				map(objects.volumes, vId => {
+					return this.docker
+						.getVolume(vId)
+						.remove()
+						.catch(e => log.error(`Unable to delete volume ${vId}`, e));
+				}),
+			)
+			.then(() => {
+				// Remove any local mode state added to the database.
+				return this.db
+					.models('app')
+					.del()
+					.where({ source: 'local' })
+					.catch(e =>
+						log.error('Cannot delete local app entries in the database', e),
+					);
 			});
+	}
```

Comment:
Please use `async/await`.

Reply:
So, in this method, I kept `.catch()` constructions. Here's why.

I wanted to catch every individual object deletion failure, so that we do attempt all objects deletion instead of stopping on the first failure.
Adding `try/catch` in every close there would be too ugly to my taste.

The last db operation catch could be done with `try/catch`, but I kept the promise-like approach for consistency with the rest of the code in this method.

Reply:
Fair enough, the code is clear, so I'm happy.